ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CompletionDeduplicationInfoIT.scala

```diff
@@ -0,0 +1,194 @@
+    ): SubmitAndWaitRequest = {
+      val request = ledger.submitAndWaitRequest(party, simpleCreate(party))
+      optSubmissionId
+        .map { submissionId =>
+          request.update(_.commands.submissionId := submissionId)
+        }
+        .getOrElse(request)
+    }
+
+    override def submitRequest(
+        ledger: ParticipantTestContext,
+        request: SubmitAndWaitRequest,
+    )(implicit ec: ExecutionContext): Future[Option[Completion]] =
+      for {
+        offset <- ledger.currentEnd()
+        _ <- ledger.submitAndWait(request)
+        completion <- singleCompletionAfterOffset(ledger, offset)
+      } yield completion
+  }
+
+  case object CommandSubmissionService extends Service[SubmitRequest] {
+    override def buildRequest(
+        ledger: ParticipantTestContext,
+        party: binding.Primitive.Party,
+        optSubmissionId: Option[SubmissionId],
+    ): SubmitRequest = {
+      val request = ledger.submitRequest(party, simpleCreate(party))
+      optSubmissionId
+        .map { submissionId =>
+          request.update(_.commands.submissionId := submissionId)
+        }
+        .getOrElse(request)
+    }
+
+    override def submitRequest(
+        ledger: ParticipantTestContext,
+        request: SubmitRequest,
+    )(implicit ec: ExecutionContext): Future[Option[Completion]] =
+      for {
+        offset <- ledger.currentEnd()
+        _ <- ledger.submit(request)
+        completion <- singleCompletionAfterOffset(ledger, offset)
+      } yield completion
+  }
+
+  private def singleCompletionAfterOffset(
+      ledger: ParticipantTestContext,
+      offset: LedgerOffset,
+  ): Future[Option[Completion]] =
+    WithTimeout(5.seconds)(
+      ledger.findCompletion(ledger.completionStreamRequest(offset)())(_ => true)
+    )
+
+  private def assertSubmissionIdIsPreserved(
+      requestedSubmissionId: Ref.SubmissionId,
+      optCompletion: Option[Completion],
+  ): Unit = {
+    val submissionIdCompletion = assertDefined(optCompletion)
+    val actualSubmissionId = submissionIdCompletion.submissionId
+    assert(submissionIdCompletion.status.forall(_.code == Status.Code.OK.value()))
+    assert(
+      actualSubmissionId == requestedSubmissionId,
+      "Wrong submission ID in completion, " +
+        s"expected: $requestedSubmissionId, actual: $actualSubmissionId",
+    )
+  }
+
+  private def assertDefaultDeduplicationTimeIsReportedIfNoDeduplicationSpecified(
+      config: LedgerConfiguration,
+      optCompletion: Option[Completion],
+  ): Unit = {
+    val completion = assertDefined(optCompletion)
+    assert(completion.status.forall(_.code == Status.Code.OK.value()))
+    val actualDeduplication = completion.deduplicationPeriod
+    assert(actualDeduplication.isDeduplicationTime)
+    assert(
+      actualDeduplication.deduplicationTime == config.maxDeduplicationTime,
+      s"The deduplication $actualDeduplication " +
+        "is not the maximum deduplication time",
+    )
```

Comment:
There's no requirement on the Ledger API that we report the default deduplication period as a deduplication time. Canton, e.g., will report some offset that it derives from the max deduplication time.

A more meaningful test would look like this:
1. Get a command accepted with a very short dedup time.
2. Then submit a command with the same change id, but without a deduplication time.
3. If we get a completion before the max dedup time has elapsed after the acceptance of the first command, then the completion should be "ALREADY_EXISTS" and the reported deduplication period should start before the accepted command.

Reply:
> There's no requirement on the Ledger API that we report the default deduplication period as a deduplication time. Canton, e.g., will report some offset that it derives from the max deduplication time.

Yes, I was just thinking about this too.

> 3. If we get a completion before the max dedup time has elapsed after the acceptance of the first command
> [...] the reported deduplication period should start before the accepted command.

KV would report the max deduplication time, other ledgers some other deduplication form, so I think this test would be best implemented as ledger-specific.

At the general API level, I'd perhaps just ensure that some deduplication info is present in the completion no matter what.

Reply:
b60c75c

Reply:
> At the general API level, I'd perhaps just ensure that some deduplication ~info~ period (in some form) is present in the completion no matter what.

Yes. That makes sense too.